components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/AuthorizationParams.kt

```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.appservices.fxaclient
+
+data class AuthorizationParams(
+    val clientId: String,
+    val scopes: Array<String>,
+    val state: String,
+    val accessType: String = "online",
+    val keyParams: AuthorizationKeyParams? = null
+) {
+    fun intoMessage(): MsgTypes.AuthorizationParams {
+        var b = MsgTypes.AuthorizationParams.newBuilder().setClientId(this.clientId)
+                .setScope(this.scopes.joinToString(" "))
+                .setState(this.state)
+                .setAccessType(this.accessType)
+        if (this.keyParams != null)
+            b = b.setKeyParams(this.keyParams.intoMessage())
+        return b.build()
+    }
+
+    // Auto Generated equals and hashCode
+    // Since AuthorizationParams holds an array of scopes
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as AuthorizationParams
+
+        if (clientId != other.clientId) return false
+        if (!scopes.contentEquals(other.scopes)) return false
+        if (state != other.state) return false
+        if (accessType != other.accessType) return false
+        if (keyParams != other.keyParams) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = clientId.hashCode()
+        result = 31 * result + scopes.contentHashCode()
+        result = 31 * result + state.hashCode()
+        result = 31 * result + accessType.hashCode()
+        result = 31 * result + (keyParams?.hashCode() ?: 0)
+        return result
+    }
+}
+
+data class AuthorizationKeyParams(
+    val codeChallenge: String,
+    val codeChallengeMethod: String = "S256",
```

Comment:
We often conflate them, but `codeChallenge` and `codeChallengeMethod` don't actually have anything to do with fetching keys. These parameters are for [PKCE](https://tools.ietf.org/html/rfc7636), which allows the OAuth flow to be used by what we call "public clients"; that is, clients that cannot maintain a secret key with which to authenticate their calls to the OAuth server.

It just so happens that all of our current public clients also use the scoped-keys feature, and that anyone using the scoped-keys feature is almost certainly a public client (because we would never deliver the user's secret keys to a server-server OAuth client). So we always see these params together in practice, even though they should not be coupled together in the user's mental model of the system. One day we might add a public client that doesn't want scoped keys, but still needs to use PKCE.

So, I don't think we should bundle them into the same object here. You could make a separate `AuthorizationPKCEParams` class to hold just the PKCE-related stuff though, if that makes things easier to manage.

Reply:
Ooo That makes tons of sense. Sorry for all the confusing attempts here 😅

Reply:
> Sorry for all the confusing attempts here

FWIW it's really useful for me, to help me understand all the things that could be clearer, so I really appreciate your efforts to work through this!